File: src/content/docs/topics/cs/leetcode-150/backtracking/040-combination-sum-ii.py
def combination_sum2(candidates, target):
    pass  # TODO: implement

def _run_tests():
    r = combination_sum2([10, 1, 2, 7, 6, 1, 5], 8)
    assert sorted(map(tuple, r)) == sorted([tuple([1, 1, 6]), tuple([1, 2, 5]), tuple([1, 7]), tuple([2, 6])])
    r2 = combination_sum2([2, 5, 2, 1, 2], 5)
    assert sorted(map(tuple, r2)) == sorted([tuple([1, 2, 2]), tuple([5])])
    assert combination_sum2([1, 2], 10) == []
    assert combination_sum2([3, 3], 3) == [[3]]
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    combination_sum2([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 1, 2, 3], 15)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf combination_sum2(13 candidates, target=15): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/backtracking/040-combination-sum-ii.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func combinationSum2(_ candidates: [Int], _ target: Int) -> [[Int]] {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:e5519429379e0f2d230af1238b327c66cec643e2f8d2485503c456e1a7917fcc
    expectEqual(Solution().combinationSum2([10, 1, 2, 7, 6, 1, 5], 8), [[1, 1, 6], [1, 2, 5], [1, 7], [2, 6]], "canonical")
    expectEqual(Solution().combinationSum2([2, 5, 2, 1, 2], 5), [[1, 2, 2], [5]], "second")
    expectEqual(Solution().combinationSum2([3, 4], 2), [], "none")
    expectEqual(Solution().combinationSum2([1], 1), [[1]], "single")
    expectEqual(Solution().combinationSum2([1, 1], 1), [[1]], "duplicate-only")
    // EXCLUDED_VECTOR nonpositive: [[0,1],1] | Candidates are positive integers.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
